plugins/dell-dock/fu-dell-dock-usb-usb4.c

```diff
@@ -0,0 +1,654 @@
+		g_prefix_error (error, "NVM Version Read Failed");
+		return NULL;
+	}
+	nvm_version = g_strdup_printf ("%02x.%02x", buf[10], buf[9]);
+	return nvm_version;
+}
+
+static gboolean
+fu_dell_dock_usb4_write_fw (FuDevice *device, FuFirmware *firmware,
+			    FwupdInstallFlags flags, GError **error)
+{
+	FuDellDockUsb4 *self = FU_DELL_DOCK_USB4 (device);
+	guint32 header_offset = 0;
+	gsize image_size = 0;
+	const guint8 *buffer;
+	g_autoptr(GTimer) timer = g_timer_new ();
+	g_autofree gchar *dynamic_version    = NULL;
+	g_autofree gchar *dynamic_vendor_id  = NULL;
+	g_autofree gchar *dynamic_product_id = NULL;
+	g_autoptr(GBytes) fw = NULL;
+
+	g_return_val_if_fail (device != NULL, FALSE);
+	g_return_val_if_fail (FU_IS_FIRMWARE (firmware), FALSE);
+
+	/* get default image */
+	fw = fu_firmware_get_bytes (firmware, error);
+	if (fw == NULL)
+		return FALSE;
+	buffer = g_bytes_get_data (fw, &image_size);
+	g_debug ("Total Image size: %" G_GSIZE_FORMAT, image_size);
+
+	/* get image version */
+	dynamic_version = g_strdup_printf ("%02x.%02x",
+					   buffer[self->blob_major_offset],
+					   buffer[self->blob_minor_offset]);
+	g_debug ("writing Thunderbolt firmware version %s", dynamic_version);
+
+	/* get image vendor-id */
+	dynamic_vendor_id = g_strdup_printf ("0x%02x%02x",
+					     buffer[self->vid_offset+1],
+					     buffer[self->vid_offset]);
+	g_debug ("writing Thunderbolt vendor-id %s", dynamic_vendor_id);
+
+	/* get image product-id */
+	dynamic_product_id = g_strdup_printf ("0x%02x%02x",
+					       buffer[self->pid_offset+1],
+					       buffer[self->pid_offset]);
+	g_debug ("writing Thunderbolt product-id %s", dynamic_product_id);
+
+	/* check header size */
+	memcpy (&header_offset, buffer, sizeof (guint32));
+	g_debug ("Header size: %" G_GUINT32_FORMAT, header_offset);
+	if (header_offset > image_size) {
+		g_set_error (error, FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE,
+			     "Image header is too big: %" G_GUINT32_FORMAT,
+			     header_offset);
+		return FALSE;
+	}
+	buffer += header_offset;
+	image_size -= header_offset;
+
+	/* firmware install */
+	if (!fu_dell_dock_usb4_hub_nvm_write(device, buffer, image_size, 0,
+					     error))
+		return FALSE;
+
+	/* dock will reboot to re-read; this is to appease the daemon */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+	fu_device_set_version_format (device, FWUPD_VERSION_FORMAT_PAIR);
+	fu_device_set_version (device, dynamic_version);
+
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_open (FuDevice *device, GError **error)
+{
+	/* FuUsbDevice->open */
+	if (!FU_DEVICE_CLASS (fu_dell_dock_usb4_parent_class)->open (device, error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_setup (FuDevice *device, GError **error)
+{
+	const gchar *version;
+
+	/* set version from USB NVM */
+	version = fu_dell_dock_usb4_show_nvm_info(device, error);
+	if (version != NULL){
```

Comment:
I think you're throwing away an error scenario here.  If `version` is `NULL`, shouldn't `*error` be set and you return `FALSE`?